src/plugins/data_source/server/data_source_service.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { first } from 'rxjs/operators';
+import { Client } from '@opensearch-project/opensearch';
+import LRUCache from 'lru-cache';
+import { Logger, PluginInitializerContext } from '../../../../src/core/server';
+import { DataSourcePluginConfigType } from '../config';
+import { IDataSourceService } from './types';
+
+export interface DataSourceServiceSetup {
+  getClientFromPool: (id: string) => Client | undefined;
+  addClientToPool: (endpoint: string, client: Client) => void;
+}
+export class DataSourceService implements IDataSourceService {
+  private openSearchClientPool?: LRUCache<string, Client>;
+  private isClosed = false;
+
+  constructor(
+    private logger: Logger,
+    private initializerContext: PluginInitializerContext<DataSourcePluginConfigType>
+  ) {}
+
+  public async setup() {
+    const config$ = this.initializerContext.config.create<DataSourcePluginConfigType>();
+    const config: DataSourcePluginConfigType = await config$.pipe(first()).toPromise();
```

Comment:
you can pass config to setup function from plugin setup, instead of parsing config again here